hivemind/moe/server/connection_handler.py

```diff
@@ -24,59 +28,118 @@ class ConnectionHandler(mp.context.ForkProcess):
+    async def rpc_info(self, request: runtime_pb2.ExpertUID, context: P2PContext) -> runtime_pb2.ExpertInfo:
         return runtime_pb2.ExpertInfo(serialized_info=MSGPackSerializer.dumps(self.experts[request.uid].get_info()))
 
-    async def forward(self, request: runtime_pb2.ExpertRequest, context: grpc.ServicerContext):
+    class _RequestUnpacker:
+
+        __slots__ = ("uid",)
+
+        def __init__(self):
+            self.uid = None
+
+        def __call__(self, request: runtime_pb2.ExpertRequest) -> Iterable[runtime_pb2.Tensor]:
+            if self.uid is None:
+                self.uid = request.uid
+            else:
+                assert self.uid == request.uid, "Expert uids differ in one request"
+
+            return request.tensors
+
+    async def _gather_inputs(
+        self, requests: AsyncIterator[runtime_pb2.ExpertRequest], context: P2PContext
+    ) -> Tuple[str, List[torch.Tensor]]:
+        unpacker = self._RequestUnpacker()
+        inputs = await gather_from_grpc(requests, unpacker, deserialize_torch_tensor)
+        return unpacker.uid, inputs
+
+    async def _process_inputs(
+        self,
+        inputs: List[torch.Tensor],
+        pool: TaskPool,
+        schema: Union[BatchTensorDescriptor, Tuple[BatchTensorDescriptor, ...]],
+    ) -> List[runtime_pb2.Tensor]:
+        return [
+            serialize_torch_tensor(t, p.compression, allow_inplace=True)
+            for t, p in zip(await pool.submit_task(*inputs), nested_flatten(schema))
+        ]
+
+    async def rpc_forward(self, request: runtime_pb2.ExpertRequest, context: P2PContext) -> runtime_pb2.ExpertResponse:
         inputs = [deserialize_torch_tensor(tensor) for tensor in request.tensors]
-        future = self.experts[request.uid].forward_pool.submit_task(*inputs)
-        serialized_response = [
-            serialize_torch_tensor(tensor, proto.compression, allow_inplace=True)
-            for tensor, proto in zip(await future, nested_flatten(self.experts[request.uid].outputs_schema))
+        expert = self.experts[request.uid]
+        return runtime_pb2.ExpertResponse(
+            tensors=await self._process_inputs(inputs, expert.forward_pool, expert.outputs_schema)
+        )
+
+    async def rpc_forward_partial(
+        self, requests: AsyncIterator[runtime_pb2.ExpertRequest], context: P2PContext
+    ) -> AsyncIterator[runtime_pb2.ExpertRequest]:
+        uid, inputs = await self._gather_inputs(requests, context)
+        expert = self.experts[uid]
+        output_split = [
+            p
+            for t in await self._process_inputs(inputs, expert.forward_pool, expert.outputs_schema)
+            for p in split_for_streaming(t, DEFAULT_MAX_MSG_SIZE // 2)
         ]
 
-        return runtime_pb2.ExpertResponse(tensors=serialized_response)
+        async for part in as_aiter(*output_split):
+            yield runtime_pb2.ExpertResponse(
+                tensors=[
+                    part,
+                ],
+            )
```

Comment:
nit:

```suggestion
            yield runtime_pb2.ExpertResponse(tensors=[part])
```

Please do the same in `rpc_backward_partial()` below :)